Source/ACE/Entity/Vendor.cs

```diff
@@ -104,53 +118,59 @@ private void ApproachVendor(Player player)
-                        WorldObject wo = WorldObjectFactory.CreateNewWorldObject(defaultItemsForSale[item.Guid].WeenieClassId);
-                        // can we stack this ?
-                        if (wo.MaxStackSize.HasValue)
+                        if ((wo.MaxStackSize.Value != 0) & (wo.MaxStackSize.Value <= fitem.Amount))
                         {
-                            if ((wo.MaxStackSize.Value != 0) & (wo.MaxStackSize.Value <= item.Amount))
-                            {
-                                item.Amount -= wo.MaxStackSize.Value;
-                                goldcost += (int)defaultItemsForSale[item.Guid].Value.Value * wo.MaxStackSize.Value;
-                                wo.StackSize = wo.MaxStackSize.Value;
-                                purchaselist.Add(wo);
-                            }
-                            // else we cant stack it or its less then max stack size.
-                            else
-                            {
-                                if (item.Amount > 0)
-                                    goldcost += (int)defaultItemsForSale[item.Guid].Value.Value * item.Amount;
-                                else
-                                {
-                                    item.Amount = 0;
-                                    goldcost += (int)defaultItemsForSale[item.Guid].Value.Value;
-                                }
-                                wo.StackSize = (ushort)item.Amount;
-                                purchaselist.Add(wo);
-                            }
+                            wo.StackSize = wo.MaxStackSize.Value;
+                            purchaselist.Add(wo);
                         }
+                        // else we cant stack but its not a single item.
                         else
                         {
-                            // single item with no stack options.
-                            item.Amount = 0;
-                            wo.StackSize = 0;
-                            goldcost += (int)defaultItemsForSale[item.Guid].Value.Value;
+                            if (fitem.Amount > 0)
+                                goldcost = goldcost + ((uint)wo.Value * fitem.Amount);
+                            else
+                            {
+                                fitem.Amount = 0;
+                                goldcost = goldcost + (uint)wo.Value;
+                            }
+                            wo.StackSize = (ushort)fitem.Amount;
                             purchaselist.Add(wo);
                         }
                     }
+                    else
+                    {
+                        // single item with no stack options.
+                        fitem.Amount = 0;
+                        wo.StackSize = null;
+                        goldcost = goldcost  + (uint)wo.Value;
+                        purchaselist.Add(wo);
+                    }
                 }
             }
 
```

Comment:
Gold cost should be scaled by the Vendor's Double property (PropertyDouble.SellPrice / 38) for their sell rate.